src/fhirtypes/ElementDefinition.ts

```diff
@@ -414,6 +416,101 @@ export class ElementDefinition {
       .join('.');
   }
 
+  /**
+   * Apply the AddElementRule to this new element using the appropriate methods
+   * for specific rules for the AddElementRule's implied rules (i.e., cardinality,
+   * type constraints, and flags).
+   * @param {AddElementRule} rule - specific instance of the rule
+   * @param {Fishable} fisher - A fishable implementation for finding definitions and metadata
+   */
+  applyAddElementRule(rule: AddElementRule, fisher: Fishable): void {
+    // The constrainType() method applies a type constraint to an existing
+    // ElementDefinition.type. Since this is a new ElementDefinition, it
+    // does not yet have a 'type', so we need to assign an "initial" value
+    // that the constrainType() method can process.
+    this.type = this.initializeElementType(rule, fisher);
+    const target = this.structDef.getReferenceName(rule.path, this);
+    this.constrainType(rule, fisher, target);
```

Comment:
I think I'm having a little trouble understanding the difference between what `initializeElementType` does and `constrainType` does...  In other words, what does `constrainType` do that `initializeElementType` did not already do?

Reply:
OK.  I've thought about this some more and have gone back and forth -- and I think your initial inclination was right, but the execution isn't quite there yet.  The `constrainType` function is very specific to the use case of constraining a broad set of types to a narrower set of types.  So your `initializeElementType` should just establish that first *broad* set of types.  I think that right now, you're maybe setting it up too narrowly.

What we really want is for the initial set of types just to represent the set of base types that the more specific types will either match exactly or match as a profile.  For example, if the types in AddElement are:
* MyCustomLogicalModel
* MyPatientProfileA (a profile on Patient)
* MyPatientProfileB (a profile on Patient)
* Reference(Practitioner)
* Reference(Group)
Then you want an initial type array of:
```
[
  { code: 'MyCustomLogicalModel' },
  { code: 'Patient' },
  { code: 'Reference' }
]
```
because that is the most basic set of types that the incoming types can all cleanly map to.

I think you can implement this by looping through the types:
* if the type `isReference`, add `{ code: 'Reference' }` to the initial types
* else `fshForMetadata` on the type and add `{ code: meta.sdType }` to the initial types
* either avoiding duplicates along the way or removing duplicates at the end

Then after that, I think `constrainType` should work as expected with the types from the `addElementRule`.